session/txn.go

```diff
@@ -294,20 +294,23 @@ func (st *TxnState) KeysNeedToLock() ([]kv.Key, error) {
 	keys := make([]kv.Key, 0, st.countHint())
 	buf := st.Transaction.GetMemBuffer()
 	buf.InspectStage(st.stagingHandle, func(k kv.Key, flags kv.KeyFlags, v []byte) {
-		if !keyNeedToLock(k, v) {
+		if !keyNeedToLock(k, v, flags) {
 			return
 		}
 		keys = append(keys, k)
 	})
 	return keys, nil
 }
 
-func keyNeedToLock(k, v []byte) bool {
+func keyNeedToLock(k, v []byte, flags kv.KeyFlags) bool {
 	isTableKey := bytes.HasPrefix(k, tablecodec.TablePrefix())
 	if !isTableKey {
 		// meta key always need to lock.
 		return true
 	}
+	if flags.HasPresumeKeyNotExists() {
```

Comment:
Should we use `flags.HasNeedCheckExists()` as it's statement level flags.

Reply:
If the flag has `PresumeKeyNotExists` but doesn't has `NeedCheckExists` flag, the key must have been locked already, locking it again doesn't matter.